CotEditor/Sources/Script.swift

```diff
@@ -27,92 +27,215 @@
+        }
+        
+        let ordering: Int?
+        if let range = name.range(of: "^[0-9]+\\)", options: .regularExpression) {
+            // Remove the parenthesis at last
+            let orderingString = name.substring(to: name.index(before: range.upperBound))
+            ordering = Int(orderingString)
+            
+            // Remove the ordering number from the script name
+            name.removeSubrange(range)
+        } else {
+            ordering = nil
+        }
+        
+        // Extract from Info.plist
+        
+        let info = NSDictionary(contentsOf: url.appendingPathComponent("Contents/Info.plist"))
+        
+        let executionModel: ScriptingExecutionModel
+        if let name = info?["CotEditorExecutionModel"] as? String {
+            executionModel = ScriptingExecutionModel(rawValue: name) ?? .unrestricted
+        } else {
+            executionModel = .unrestricted
+        }
+        
+        let eventTypes: [ScriptingEventType]
+        if let names = info?["CotEditorHandlers"] as? [String] {
+            eventTypes = names.flatMap { ScriptingEventType(rawValue: $0) }
+        } else {
+            eventTypes = []
         }
         
-        NSWorkspace.shared().activateFileViewerSelecting([self.url])
+        self.init(url: url, name: name, type: type, executionModel: executionModel, eventTypes: eventTypes, shortcut: shortcut, ordering: ordering)
     }
     
     
+    /// Create and return an user script instance
+    ///
+    /// - returns: An instance of `Script` created by the receiver.
+    ///            Returns `nil` if the script type is unsupported.
+    func makeScript() -> Script? {
+        switch self.type {
+        case .appleScript:
+            switch self.executionModel {
+            case .unrestricted: return UnrestrictedAppleScript(with: self)
+            case .persistent: return PersistentOSAScript(with: self)
+            }
+        case .shellScript: return ShellScript(with: self)
+        case .unknown: return nil
+        }
+    }
+}
+
+
+
+protocol Script {
     
-    // MARK: Private Methods
+    // MARK: Properties
     
-    /// append message to console panel and show it
-    fileprivate static func writeToConsole(message: String, scriptName: String) {
-        
-        DispatchQueue.main.async {
-            ConsolePanelController.shared.showWindow(nil)
-            ConsolePanelController.shared.append(message: message, title: scriptName)
-        }
+    /// A script descriptor the receiver was created from.
+    var descriptor: ScriptDescriptor { get }
+    
+    // MARK: Methods
+    
+    /// Execute the script with the default way.
+    func run() throws
+    
+    
+    /// Execute the script by sending it the given Apple event.
+    ///
+    /// Events the script cannot handle must be ignored with no errors.
+    func run(withAppleEvent event: NSAppleEventDescriptor?) throws
```

Comment:
As long as `Script` protocol is also used with non-AppleScript objects, it shouldn't have AppleEvent specific methods.